### ISSUE TITLE
Remove InsertAfter and InsertBefore

### ISSUE THREAD

**Issue by andreasohlund** [Particular/NServiceBus#3582]
All usages we've seen so far has turned out to be code smells/bad ideas.

Did some research:

InsertBeforeIfExists: only used by recoverability (we're about to merge those soon)
InsertBefore: used by Encryption and DataBus to go before incoming mutators. We can move mutators to the connector instead

InsertAfterIfExists: SubscriptionReceiverBehavior to go after UOW (no idea why this is, think we can remove this)
InsertAfter: Encryption and Databus to go after mutators on the outgoing pipe => we can move mutators to the connector instead

In short: How about we obsoleted them?

**Comment by danielmarbach** [comment]
I'm game!

**Comment by timbussmann** [comment]
I think we currently still need them, e.g. for emulating the custom FaultManager abilities in v5.

**Comment by andreasohlund** [comment]
Hmm, I thought just inserting into the physical receive stage was good enough?

**Comment by timbussmann** [comment]
nope, it needs to be before the MoveFaultToErrorQueueBehavior and dependent of usage scenario between FLR & SLR or after SLR.

**Comment by andreasohlund** [comment]
Hmm, is the code somewhere? (don't get why it needs to go before move faults to error?)

**Comment by timbussmann** [comment]
@andreasohlund https://github.com/Particular/docs.particular.net/pull/1300/files#diff-fb55e3b39288b3167efeeb42f6f9e58fR40

**Comment by mikeminutillo** [comment]
Is there a deterministic order for behaviors within a stage? i.e. Will the pipeline always end up in the same order? If not, we run the risk that it works 100% of the time on our machines and on our build server and slightly re-orders the behaviors in a prod environment giving us weird results.

**Comment by timbussmann** [comment]
> If not, we run the risk that it works 100% of the time on our machines and on our build server and slightly re-orders the behaviors in a prod environment giving us weird results.

I think the idea is to design behaviors in a way that the order within a specific stage doesn't matter.

> Is there a deterministic order for behaviors within a stage?

we don't provide any guarantees. So it would be wrong to assume a certain place.

**Comment by bording** [comment]
I had been wondering the same thing as @mikeminutillo. While we assume that behaviors are being written to be order-agnostic, do we have a way to actually verify that we haven't accidentally introduced an implicit order dependency?

From what I've seen of the pipeline implementation, while we may not guarantee any ordering, in practice you're likely to see the same order, so you might not even be aware that you've written a behavior that inadvertently has an order dependency.

**Comment by timbussmann** [comment]
> do we have a way to actually verify that we haven't accidentally introduced an implicit order dependency?

not within the same stage. How could we do that?

**Comment by danielmarbach** [comment]
Well if we have, we'll sooner or later find out with our test suite. As far
as we can say we haven't detected an issue and therefore the problem
doesn't exist ;)

**Comment by bording** [comment]
> not within the same stage. How could we do that?

The only thing that comes to mind would be to somehow put the pipeline creation into a special test mode that would actually randomize the order it adds behaviors to a given stage, guaranteeing that they're in a different order all the time.

Looking at `PipelineModelBuilder` it seems like something like that could be possible, though needing to preserve some order around the `Befores` and `Afters` makes that more complicated.

If we end up removing those, then it would be much easier.

**Comment by danielmarbach** [comment]
Aren't we overthinking this?

**Comment by andreasohlund** [comment]
After talking to @timbussmann we concluded that the statement

>All usages we've seen so far has turned out to be code smells/bad ideas.

Is true.

That said our pipeline is always evolving so **we need the InsertBefore/After as temporary workarounds** until we're able to adjust the pipeline (more stages etc).

By adjusting we mean:

* Moving things to the connectors (eg the mutators should probably be moved)
* Merging behaviors
* Removing behaviors (eg the unit of work one)
* Add more stages
* etc

### Summary

* Usage is a smell so look out for it in code reviews
* Avoid using them if possible

### Actions

* Open a PR/issue against docs to explain why insertbefore/after should be avoided (probably a new "advanced section")
*  Plan the removal of internal usages

**Comment by andreasohlund** [comment]
Docs issue raised

https://github.com/Particular/docs.particular.net/issues/2022

@danielmarbach please close this one once you have the issue for "Plan the removal of internal usages" raised

**Comment by danielmarbach** [comment]
Raised issues about InsertAfterIfExists and InsertBeforeIfExists. InsertAfter and InsertBefore are not used by the core (except for the tests which test the behavior of these methods which is fine). But the docs have a multitude of references see

https://github.com/search?utf8=%E2%9C%93&q=org%3AParticular+InsertAfter&type=Code&ref=searchresults
https://github.com/search?utf8=%E2%9C%93&q=org%3AParticular+InsertBefore&type=Code&ref=searchresults

Since we scope this issue to mostly core I'm going to close this.
